#' Conduct Bayesian phylogenetic analysis (MrBayes)
#'
#' Create a NEXUS batch file and a bash script for conducting Bayesian
#' phylogenetic analysis via MrBayes.
#'
#' @param mb_dir Path to the target directory for the MrBayes analysis, which
#'   should contain the multiple sequence alignment as a NEXUS file.
#' @param aln_file Name of the NEXUS file containing the multiple sequence
#'   alignment.
#' @param mbparts_obj A vector representing MrBayes partition definitions, as
#'   generated by `get_pfresults()`.
#' @param pf_results_obj A table containing the main results of the
#'   corresponding PartitionFinder analysis, as generated by `get_pfresults()`.
#' @param ngen An integer specifying the number of generations for the analysis.
#' @param bscr_dir Path to the directory in which the MrBayes bash script will
#'   be created.
#' @param run If TRUE, the bash script is run in Windows Subsystem for Linux
#'   (WSL) as a native R process.
#' @return No objects are returned. A NEXUS batch file named `mrbayes.nex`
#'   (containing partition definitions, nucleotide substitution model
#'   specifications and analysis parameters) and a bash script named
#'   `mrbayes.sh` are created in the specified script directory. If run, the
#'   bash script performs the MrBayes analysis using the batch file, with
#'   results saved in the specified RAxML-NG directory.
#' @examples
#' phyl_mb()
#' @export
phyl_mb <- function (
  mb_dir, aln_file, mbparts_obj, pf_results_obj, ngen, bscr_dir, run = FALSE
) {
  # split model strings by "+":
  pf_mods <- stringr::str_split(pf_results_obj$'Best Model', "\\+")
  
  # Create batch file, specifying partitioning scheme, substitution models, run
  # parameters, etc., to run MrBayes using the alignment NEXUS file (see MrBayes
  # manual for formatting):
  readr::write_lines(  # write the following lines...
    c(
      "begin mrbayes;",
      paste0("execute ",aln_file,";"),
      mbparts_obj,  # specify charsets (NB -- MrBayes skips NEXUS 'sets' block)
      paste0(
        "partition pf = ", nrow(pf_results_obj), ": ",
        paste(pf_results_obj$'Partition names', collapse = ", "), ";"
      ),
      "set partition = pf;"
    ),
    # ...to the following file:
    paste0(mb_dir,"mrbayes.nex")
  )
  
  # ~ nucleotide substitution model specification:
  purrr::walk2(  # for each best fit model and partition number,
    pf_mods, seq_len(nrow(pf_results_obj)), ~ {
      nst <-  # specify no. of substitution types (1, 2, 6 or 'mixed')
        if (any(stringr::str_detect(.x, "JC|JC69|F81"))) { 1
        } else if (any(stringr::str_detect(.x, "K80|K2P|HKY|HKY85"))) { 2
        } else if (any(stringr::str_detect(.x, "GTR"))) { 6
        } else { "mixed" }
      rates <-  # specify among-site rate variation model
        dplyr::case_when(
          "G" %in% .x & "I" %in% .x ~ "invgamma",
          "G" %in% .x & !("I" %in% .x) ~ "gamma",
          !("G" %in% .x) & "I" %in% .x ~ "propinv",
          !("G" %in% .x) & !("I" %in% .x) ~"equal"
        )
      # **append** lines to apply relevant nst and rate parameters:
      readr::write_lines(  # append the following lines...
        c(  # apply relevant parameters to partition:
          paste0("lset applyto=(", .y, ") nst=", nst, ";"),  
          paste0("lset applyto=(", .y, ") rates=", rates, ";"),
          # no. gamma rate categories (4 is reasonable approximation):
          paste0("lset applyto=(", .y, ") ngammacat=4;")
        ),
        # ...to the batch file:
        paste0(mb_dir,"mrbayes.nex"), append = TRUE
      )
    }
  )
  
  # ~ remaining parameters:
  readr::write_lines(  # **append** the following lines...
    c(
      # ensure all partitions have different parameters:
      "unlink statefreq=(all) revmat=(all) shape=(all) pinvar=(all);",
      # # (NB -- following line required??? Sometimes results in error)
      # "prset applyto=(all) ratepr=variable;",  # allow different rates
      
      # MCMC chain parameters:
      # (NB -- split mcmcp over multiple lines for readability purposes only)
      paste0("mcmcp ngen=",format(ngen, scientific = FALSE),";"),
      # (prevent scientific formatting of integer)
      "mcmcp samplefreq=100 printfreq=1000 diagnfreq=1000;",
      # (default samplefreq=100 works well for most analyses)
      "mcmcp nchains=4 nruns=2 savebrlens=yes temp=0.2;",
      
      "mcmc relburnin=yes burninfrac=0.25;",  # run analysis
      # (use e.g. 'relburnin=no burnin=250000' for specific burnin)
      
      "sump;", # summarise resulting parameter values
      "sumt relburnin=yes burninfrac=0.25 contype=allcompat conformat=simple;",
      # conformat=simple allows reading into R with posterior prob.s;
      # change to conformat=figtree to allow reading into Figtree
      "end;"
    ),
    # ...to the batch file:
    paste0(mb_dir,"mrbayes.nex"), append = TRUE
  )
  
  # create bash script for running MrBayes:
  readr::write_lines(  # write the following lines...
    c(
      "#!/bin/bash", # bash script header
      "# Bayesian phylogenetic tree estimation (MrBayes)",
      paste0('cd "',mb_dir,'"'), # change directory
      "mb-mpi mrbayes.nex" # run MrBayes analysis
      # ('mb' = sequential version, 'mb-mpi' = parallel version)
    ),
    # ...to the following file:
    paste0(bscr_dir,"mrbayes.sh")
  )
  
  if (run) {  # if specified, run the script via WSL:
    system(paste0('bash -c "',bscr_dir,'mrbayes.sh"'))
  }
  
  # Check output (generated by 'sump' and 'sumt') to ensure convergence:
  # standard deviation of split frequencies should be <0.01,
  # potential scale reduction factor (PSRF) values should be ~1,
  # estimated sample sizes (ESS) should be >100.
  
  # Open both trace (.p) files in Tracer, and for each one (and combined),
  # examine trace for every parameter.
  
  # To run 'sumt' and 'sump' on old files using mrbayes:
  # run 'execute aln.nex' (to read in matrix) before 'sumt filename=aln.nex outputname=DELETETHIS';
  # run 'sump filename=aln.nex'.
  # (in both cases, press 'n' when prompted to overwrite existing output)
}