#' Test nucleotide substitution models (PartitionFinder)
#'
#' Create a configuration file and bash script for determining the best-fit
#' partitioning scheme and nucleotide substitution model(s) for a multiple
#' sequence alignment using PartitionFinder.
#'
#' @param pf_dir Path to the target directory for the PartitionFinder analysis,
#'   which should contain the multiple sequence alignment as a PHYLIP file.
#' @param aln_file Name of the file containing the PHYLIP multiple sequence
#'   alignment to be analysed by PartitionFinder.
#' @param cpart Is the alignment to be partitioned by codon position?
#' @param cpos_obj A list of vectors describing the positions of codons 1, 2 and
#'   3 in the alignment, as generated by `get_cmasks()`.
#' @param marker For a single-marker (not concatenated) analysis, the name of
#'   the genetic marker.
#' @param aln_obj For a single-marker (not concatenated) analysis, a `phangorn`
#'   phyDat object containing the multiple sequence alignment.
#' @param conc_alns_obj For concatenated sequences, a list of `Biostrings`
#'   DNAStringSet objects containing alignments corresponding to the
#'   concatenated markers.
#' @param bscr_dir Path to the directory in which the PartitionFinder bash
#'   script will be created.
#' @param run If TRUE, the bash script is run in Windows Subsystem for Linux
#'   (WSL) as a native R process.
#' @return No objects are returned. A configuration file named
#'   `partition_finder.cfg` and a bash script named `partitionfinder.sh` are
#'   created in the specified PartitionFinder and script directories,
#'   respectively. If run, the bash script generates results in a sub-directory
#'   `./analysis` within the PartitionFinder directory.
#' @examples
#' testmods_pf()
#' @export
testmods_pf <- function (
  pf_dir, aln_file, cpart = FALSE, cpos_obj = NULL,
  marker, aln_obj, conc_alns_obj = NULL, bscr_dir, run = FALSE
) {
  # create the PartitionFinder configuration file:
  readr::write_lines(  # output the following lines...
    c(
      paste0("alignment = ",aln_file,";"),  # alignment file name
      "branchlengths = unlinked;",
      "models = all;",
      "model_selection = aicc;",
      "[data_blocks]",  # start data blocks definition section
      
      # lines specifying data blocks:
      if (cpart) {  # if partitioning by codon (including concatenated seqs),
        purrr::map2(cpos_obj, names(cpos_obj), ~ {  # for each block,
          # string of multiple sites separated by spaces:
          paste0(.y," = ",paste(.x, collapse = " "),";")
        })
      } else if (marker == "conc") {  # otherwise, for concatenated alignments,
        purrr::map2(conc_alns_obj, names(conc_alns_obj), ~ {
          # for each single-marker alignment,
          paste0(.y," = ",starts[.y],"-",ends[.y],";")
        })
      } else {  # otherwise (i.e. for single, unpartitioned marker),
        paste0(marker," = 1-",ncol(phangorn::phyDat2MultipleAlignment(aln_obj)),";")
      },
      
      # remaining lines:
      "[schemes]",  # partitioning scheme section
      "search = greedy;"
    ),
    # ...to the following file:
    paste0(pf_dir,"partition_finder.cfg")
  )
  
  # create bash script for running PartitionFinder:
  readr::write_lines(  # output the following lines...
    c(
      "#!/bin/bash",  # bash script header
      "# Selecting best-fit partitioning scheme and substitution models (PartitionFinder)",
      paste0('cd "',pf_dir,'"'),  # change directory
      "rm -R analysis/",  # remove results of previous analysis
      "rm log.txt",  # remove log file (otherwise current log is just appended)
      paste0(  # run PartitionFinder analysis
        "python ~/partitionfinder-2.1.1/PartitionFinder.py ./ ")
      # "--no-ml-tree")  # (NB -- only if cannot remove unused sites from alignment;
      # but this uses Maximum Parsimony, rather than Maximum Likelihood)
    ),
    # ...to the following file:
    paste0(bscr_dir,"partitionfinder.sh")
  )
  
  if (run) {  # if specified, run the script via WSL:
    system(paste0('bash -c "',bscr_dir,'partitionfinder.sh"'))
  }
}